### ISSUE TITLE
cannot run composer install if from git repo

### ISSUE THREAD

**Issue by vesper8** [composer/composer#6703]
I have a very weird issue it seems

I one of my sites and only only.. I've suddenly lost the ability to run composer install without a previously working composer.lock file

```
...
  Problem 10
    - The requested package maddons/laravel_55 dev-master exists as maddons/laravel_55[dev-develop] but these are rejected by your constraint.
  Problem 11
    - The requested package maddons/mailing dev-master exists as maddons/mailing[dev-develop] but these are rejected by your constraint.
  Problem 12
    - The requested package maddons/mapping dev-master exists as maddons/mapping[dev-develop] but these are rejected by your constraint.
...
  Problem 21
    - The requested package maddons/laravel_55_dev dev-master exists as maddons/laravel_55_dev[dev-develop] but these are rejected by your constraint.
  Problem 22
    - The requested package maddons/special_dev dev-master exists as maddons/special_dev[dev-develop] but these are rejected by your constraint.
  Problem 23
    - The requested package tools/modulator dev-master exists as tools/modulator[dev-develop] but these are rejected by your constraint.
```

Nearly every package I load is a custom package which in turns loads packages of its own. I've been using a package grouping system like this for a long time and never had an issue.

Recently I decided to start using 3 branches on this particular repo. develop, staging and master

I wonder if the name of my branch "develop" is causing a problem here?

all my composer.json files (the main one and the ones in the custom local packages) all have this:

```
    "config": {
        "preferred-install": "dist",
        "sort-packages": true,
        "optimize-autoloader": true
    },
    "minimum-stability": "dev",
    "prefer-stable": true,
```

After scratching my head for so long.. I realized that if I delete the .git folder, the problem goes away!!? I wasn't able to find any mention of this particular problem.. so I hope someone can help me because I'm pulling my hair over this

**Comment by alcohol** [comment]
Need a simplified reproducible example.

**Comment by vesper8** [comment]
I was worried I wouldn't be able to reproduce this with a simple example.. but thankfully I just did!! So hopefully you can help me

I made this extremely simple repo
https://github.com/vesper8/composer-test

it has a master branch and a develop branch. Both branches contain the same 3 files
a main composer.json which loads a single local package which has its own composer.json
and a .gitignore file

If you clone the master branch you should be able to run 'composer install' no problem

If you switch to the develop branch and try to run the same composer install you get the problem:

```
 ✝ ~/composer-test  develop±  rm -rf vendor
 ✝ ~/composer-test  develop±  rm composer.lock
 ✝ ~/composer-test  develop  composer install
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - The requested package maddons/laravel_55 dev-master exists as maddons/laravel_55[dev-develop] but these are rejected by your constraint.

 ✘ ✝ ~/composer-test  develop
```

It's important that you rm the composer.lock as well as rm -rf the vendor folder

If you just delete the lock, but leave the vendor folder, then the problem doesn't occur

**Comment by Seldaek** [comment]
The problem is your maddons/laravel_55 project there within the repository has no version, so the `path` repository will guess the version based on git information, and as it is within the main repository when you switch branches you change its version from dev-master to dev-develop etc..

You can also use the COMPOSER_ROOT_VERSION env var to define this version, or you can hardcode a version in the composer.json of maddons/laravel_55.. I'm not sure what you are up to exactly nor what your workflow is, so not sure what solution is best, but those are your options.

**Comment by vesper8** [comment]
thank you @Seldaek you just put an end to months of annoyance and this crypt I lovingly titled wtf.sh

```
cd ~/ns.develop
rm -rf ./vendor
rm -f ./composer.lock

rm -rf ../wtf/vendor
rm -rf ../wtf/modules
rm -f ../wtf/composer.lock
rm -f ../wtf/composer.json

cp ./composer.json ../wtf/composer.json
cp -a ./modules ../wtf/
cd ~/ns/2017/wtf/
composer install

cd ~/ns.develop
cp -a ../wtf/vendor ./
cp ../wtf/composer.lock ./composer.lock
```

I used your advice and simply added 
```
"version": "dev-master",
```

in all of my "maddons' files and that did the trick nicely

If you were wondering what I'm doing, this is simply my way of organizing groups of dependencies together so I can quickly scaffold new projects. For example my own local "broadcasting" composer.json is just a group of several open source packages related to broadcasting that i commonly use. And I've got 45 of these "dependency" groups at the moment

Cheers!
